### PR TITLE
Fix bugs in ModifySafeMap

### LINKED ISSUE #14271: SEGV (on Linux) with latest Minetest Master
### Minetest version

```true
Latest master (a8cf10b0b523bf4c3234be29b0dd51e719cb00c1)
```

### Active renderer

N/A

### Irrlicht device

N/A

### Operating system and version

Linux, Fedora 38

### CPU model

Intel

### GPU model

N/A

### OpenGL version

N/A

### Summary

Here's the backtrace:
```
Thread 44 "Emerge-10" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fff4923a6c0 (LWP 29773)]
0x0000000000000000 in ?? ()
(gdb) bt
#0  0x0000000000000000 in ?? ()
#1  0x0000000000a9867f in ServerEnvironment::addActiveObjectRaw (this=this@entry=0x2b46c50, object_u=std::unique_ptr<ServerActiveObject> = {...},
set_changed=set_changed@entry=true, dtime_s=dtime_s@entry=0) at /home/lars/dev/minetest/minetest/src/serverenvironment.cpp:1889
#2  0x0000000000a98ce9 in ServerEnvironment::addActiveObject (this=this@entry=0x2b46c50, object=std::unique_ptr<ServerActiveObject> = {...})
at /usr/include/c++/13/bits/unique_ptr.h:197
#3  0x0000000000831bb1 in ModApiEnv::l_add_entity (L=0x7fffd7fe0380, L@entry=<error reading variable: value has been optimized out>)
at /usr/include/c++/13/bits/unique_ptr.h:197
#4  0x00000000007fcfca in script_exception_wrapper (L=0x7fffd7fe0380, f=<optimized out>)
at /home/lars/dev/minetest/minetest/src/script/common/c_internal.cpp:41
#5  0x00007ffff7943af0 in lj_BC_FUNCCW () from /lib64/libluajit-5.1.so.2
#6  0x00007ffff7956be0 in lua_pcall () from /lib64/libluajit-5.1.so.2
#7  0x000000000081637a in ScriptApiNode::node_on_construct (this=0x138b7c8, p=..., node=...)
at /home/lars/dev/minetest/minetest/src/script/cpp_api/s_node.cpp:171
#8  0x0000000000a97d8d in ServerEnvironment::setNode (this=this@entry=0x2b46c50, p=..., n=...)
at /home/lars/dev/minetest/minetest/src/serverenvironment.cpp:1119
#9  0x000000000082e71a in ModApiEnv::l_set_node (L=0x7fffd7fe0380, L@entry=<error reading variable: value has been optimized out>)
at /home/lars/dev/minetest/minetest/src/script/lua_api/l_env.cpp:256
#10 0x00000000007fcfca in script_exception_wrapper (L=0x7fffd7fe0380, f=<optimized out>)
at /home/lars/dev/minetest/minetest/src/script/common/c_internal.cpp:41
#11 0x00007ffff7943af0 in lj_BC_FUNCCW () from /lib64/libluajit-5.1.so.2
#12 0x00007ffff7956be0 in lua_pcall () from /lib64/libluajit-5.1.so.2
#13 0x000000000081637a in ScriptApiNode::node_on_construct (this=0x138b7c8, p=..., node=...)
at /home/lars/dev/minetest/minetest/src/script/cpp_api/s_node.cpp:171
#14 0x0000000000a97d8d in ServerEnvironment::setNode (this=this@entry=0x2b46c50, p=..., n=...)
```

This is due to some commit AFTER b12be0498

### Steps to reproduce

Somewhat reliably with Mineclonia. Just starting a new world, after a few seconds MT crashes.

### PR DESCRIPTION
**Pull request by sfan5**
fixes #14271

### CODE REVIEW

File: src/util/container.h [hunk 1]
```diff
@@ -376,21 +376,33 @@ class ModifySafeMap
...
 			return;
 		}
-		if (m_iterating)
-			m_new.emplace(key, value);
-		else
-			m_values.emplace(key, value);
+		if (m_iterating) {
+			auto it = m_values.find(key);
+			if (it != m_values.end()) {
+				it->second = V();
```
Comment (lhofhansl) [inline src/util/container.h:382]: Sorry I am missing something. I'm sure there's a good reason, but why do we need to set ->second to a new value when we will set right below?

Comment (sfan5) [reply src/util/container.h:382]: This modifies `m_values`, below is `m_new`

Comment (savilli) [inline src/util/container.h:382]: Can't we do just
```cpp
it->second = value;
return;
```
here?

Comment (sfan5) [reply src/util/container.h:382]: That's possible and works. But in the context of our usecase I think it makes sense that all kind of new values only show up after the iteration is finished.

Comment (savilli) [reply src/util/container.h:382]: Makes sense

**Comment by lhofhansl** [comment]
Works.

**Review by SmallJoker — APPROVED** [review]
No segfault & tests pass. Looks good.

**Pull request merged**